### PR TITLE
[DistillationTrainer refactor] Accept a `prompt` column alongside `messages`

### PR DESCRIPTION
**Pull request by qgallouedec**
*Plan step 3.1 — stacked on PR #6460. First PR of Phase 3 (dataset contract → prompt-only). Part of #6449.*

The trainer currently only accepts a `messages` (conversational language-modeling) dataset. This PR additively teaches the collator the forward-looking **prompt-only** format — a `prompt` column of conversational messages, the same column GRPO consumes — as the first step toward making prompt-only the sole contract (later in Phase 3).

### Trainer
- `_DistillationCollator` now branches on the example's columns: a `prompt` column is treated as a completion-less prompt (the student generates on-policy), rendered and tokenized through the exact same path as the existing `messages` prompt (`apply_chat_template(..., add_generation_prompt=True)`, prompt-budget truncation, every label masked). A `messages` column keeps its current prompt/last-assistant-turn split. Nothing about the `messages` path changes.
- `prompt` is registered in `_set_signature_columns_if_needed` (moving toward GRPO's pinned `["prompt", "image", "images"]`, which lands at plan step 3.6).

### Scope
Conversational `prompt` (a list of messages) only — the direct analog of the existing conversational `messages` support. Standard (plain-string) prompts arrive with the GRPO `_tokenize_prompts` adoption in Phase 5.

### Tests
- `test_collator_accepts_prompt_only_dataset`: a `conversational_prompt_only` batch collates to a fully-masked, prompt-only tensor set (`labels == -100`, `input_ids == prompts`).
- `test_train_runs_with_prompt_only_dataset`: end-to-end prompt-only training, added as a **strict xfail**. Prompt-only datasets carry no dataset-completion tokens, so `num_items_in_batch` (counted from the raw dataloader labels *before* on-policy generation replaces them) is `0`, making the loss `sum / 0 = NaN`. This is the same denominator bug already documented at plan step 0.3; both un-xfail together when it is fixed at plan step 5.6.

### Verification (pytest)

Full `test_distillation_trainer.py`: **38 passed, 2 xfailed** (the pre-existing `num_items_in_batch` xfail + the new prompt-only one). Collator test passes; the training xfail deterministically fails on the NaN loss.

### Risk

Low. Purely additive — the `messages` path is byte-for-byte unchanged, and the new `prompt` branch reuses the same render/tokenize code. The only new runtime surface is column detection at the top of the collator loop.

---

> [!NOTE]
> **Low Risk**
> Additive column branching in the collator; the `messages` code path is untouched and prompt-only rows reuse the same tokenization path.
> 
> **Overview**
> **DistillationTrainer** can now ingest GRPO-style **prompt-only** datasets: examples with a `prompt` column (conversational messages, no assistant completion in the data) are collated like prompt-only `messages` rows—all labels masked, completions left for on-policy generation. The existing **`messages`** path (prompt = everything before the last assistant turn) is unchanged.
> 
> `_DistillationCollator` detects `prompt` vs `messages` at the start of each example; docs describe both contracts. **`prompt`** is added to `_set_signature_columns_if_needed` so the column is not stripped by the trainer.
> 
> Tests add **`test_train_runs_with_prompt_only_dataset`** as a **strict xfail**: with no dataset completion tokens, `num_items_in_batch` stays 0 before generation and the loss becomes NaN—the same denominator issue as the existing on-policy `num_items_in_batch` test, slated for plan step 5.6.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 71883b9d39be997dc5f61c23c69a8f72addc25d4. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/distillation/distillation_trainer.py [hunk 1]
```diff
@@ -176,11 +182,16 @@ def __call__(self, examples: list[dict[str, Any]]) -> dict[str, torch.Tensor]:
...
-            messages = example[self.messages_key]
-
-            # Split: prompt = everything before the last assistant turn, completion = last assistant turn
-            has_completion = len(messages) > 1 and messages[-1].get("role") == "assistant"
-            prompt_messages = messages[:-1] if has_completion else messages
+            if "prompt" in example:
+                # Prompt-only dataset (the forward-looking format, shared with GRPO): no completion to train on, the
+                # student generates one on-policy.
+                prompt_messages = example["prompt"]
+                has_completion = False
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py:189]: Intentional; item 12 is the plumbing step; the prompt-only training path is strict-xfail'd to document the zero-denominator gap, and the fix (compute num_items from the generated completions) is the very next PR, item 14 in #6449. Advanced users can pass a prompt column now; correct training lands one PR later.

File: tests/experimental/test_distillation_trainer.py [hunk 2]
```diff
@@ -266,6 +266,70 @@ def test_train_updates_params(self):
...
+            model=self.model_id,
+            teacher_model=self.model_id,
+            args=self._make_args(max_steps=1, learning_rate=0.1),
+            train_dataset=dataset,
+            processing_class=self.tokenizer,
+        )
+
+        trainer.train()
+
+        assert all(torch.isfinite(param).all() for param in trainer.model.parameters())
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply tests/experimental/test_distillation_trainer.py:296]: Already fixed on the current tip, there's a single `test_train_runs_with_prompt_only_dataset` now (the duplicate was on an earlier commit).

File: trl/experimental/distillation/distillation_trainer.py [hunk 3]
```diff
@@ -176,11 +182,16 @@ def __call__(self, examples: list[dict[str, Any]]) -> dict[str, torch.Tensor]:
...
+            else:
+                messages = example[self.messages_key]
+                # Split: prompt = everything before the last assistant turn, completion = last assistant turn
+                has_completion = len(messages) > 1 and messages[-1].get("role") == "assistant"
+                prompt_messages = messages[:-1] if has_completion else messages
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py:194]: IWOPDTrainer is a deliberately frozen snapshot (flagged in its docstring), it doesn't track the base trainer's later changes; the alignment rule is for actively-maintained duplicates.

**Review by cmpatino — APPROVED** [review]
